### PR TITLE
[.NET] Dutch DateTimeModel initial support

### PR DESCRIPTION
**Pull request by LionbridgeCSII**
352 pass, 29 fail.

Most of the ambiguous test cases with 'voor' have been updated to be resolved as 'before' instead of 'for'. 
Exceptions are: 
- Duration patterns with 'all' (e.g. 'all year') where the 'for' interpretation seems more appropriate (e.g. 'for the whole year' vs 'before the whole year'). In this case 'voor' has been included in the Duration extractions.
- 2 points ranges (still failing) where the 'before' interpretation does not make sense e.g. 'before between 2015 and 2018' ('voor tussen 2015 en 2018').

The resolution of test cases whose meaning does not match the English counterparts has been updated and new test cases have been added to better match the English inputs. 
For example:
- "Ik ben aanstaande september weg" -> "Ik ben deze september weg" (I'll be out this September).
- "Ik zal vertrekken van aanstaande mei tot okt 2020" -> "Ik zal vertrekken van deze mei tot okt 2020" (I will leave from this May to Oct 2020).
- "Heb je een arrangement op aanstaande vrijdag?" -> "Heb je een arrangement op volgende vrijdag?" (Do you have any arrangement on next Friday?). "Aanstaande" is interpreted as "upcoming" and has a different resolution than "next".
- "Gisterenavond verdwenen 26 mensen" -> "Gisterennacht verdwenen 26 mensen" (Last night 26 people disappeared).
- "Regel een Skypegesprek van 15 minuten voor volgende maandag of dinsdag na 13:00 GMT" -> "Regel een Skypegesprek van 15 minuten volgende maandag of dinsdag na 13:00 GMT" (set up a 15 minute skype call next Monday or Tuesday after 1pm GMT.).
- "Ik herinner me  de datum niet, het zou volgende maandag of afgelopen maandag moeten zijn" -> "Ik herinner me  de datum niet, het zou volgende maandag of vorige maandag moeten zijn" (I don't remember the date, it should be next Monday or previous Monday.). "Afgelopen" is interpreted as "past" and has a different resolution than "previous".
- "Ik ga morgenavond terug" -> "Ik ga morgennacht terug" (I'll go back tomorrow night).
- "Laten we over 3 minuten vanaf vandaag beginnen" -> "Laten we 3 minuten vanaf vandaag beginnen" (Let's start 3 minutes from today).
- "Het gebeurde op vorige week dinsdag om 2 uur 's middags" -> "Het gebeurde op vorige week dinsdag tegen 2 uur 's middags" (It happened on previous Tuesday by 2 in the afternoon).
- "laten we dat ergens in de komende weken vastleggen, okay?" -> "laten we dat ergens in de komende paar weken vastleggen, okay?" (let's arrange that over the next couple weeks, ok?).
- "De deur is open vanaf het middaguur vandaag tot het middaguur morgen" -> "De deur is open vanaf middag vandaag tot ochtend morgen" (The door is opened from today pm to tomorrow am.).
- "Cortana kan ons helpen een tijdstip te vinden op maandag 12-4" -> "Cortana kan ons helpen een tijdstip te vinden maandag 12-4" (Cortana can help us find a time Monday 12-4.). "Op maandag 12-4" is interpreted as Date instead of DateTimePeriod (same as "on Monday 12-4").
- "Ik ben 4:00 tot 7 weg" -> "Ik ben weg 4:00 tot 7" (I'll be out 4:00 to 7 oclock). Patters like "4:00 tot 7" that end with a pure number are extracted only if they occur at the end of the sentence.

### CODE REVIEW

**Comment by LionbridgeCSII** [comment]
@tellarin, @iMicknl, ready for review.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (iMicknl) [inline Specs/DateTime/Dutch/DateTimeModel.json:691]: ```"Input": "Ik ga 14 oktober om 8:00:31 terug"```

@tellarin what is the resolution for other (European) locales? I think this enhancement should be logged in a feature request. In locales where they use 24 hour notation, 8:00:31 should maybe not be mapped to 20:00:31. However, '8 uur' (spoken language) should be mapped to both.

Not an issue for this PR, but maybe good to already log it as low prio case? Would be good to see feedback eventually.

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (iMicknl) [inline Patterns/Dutch/Dutch-DateTime.yaml:1278]: Good catch! Small improvement could be to add 'goeiemorgen' / 'goeiemiddag', used in Flemish (nl-be). But I can add this in a follow up PR as well.

https://www.vlaanderen.be/taaladvies/goedemorgen-goeiemorgen-goede-morgen-goeie-morgen-goedenmorgen
https://www.vlaanderen.be/taaladvies/goedemiddag-goeiemiddag-goede-middag-goeie-middag-goedenmiddag

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (iMicknl) [inline Patterns/Dutch/Dutch-DateTime.yaml:1277]:
```suggestion
    '\b(lunch)$': '(?<!\b(op|om|voor|na(ar)?|rond)\b\s)(lunch)'
```

Not sure what `ana(ar)?` should be, I assume it is 'na' (after).

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (iMicknl) [inline Patterns/Dutch/Dutch-DateTime.yaml:120]: _(todo @iMicknl: double check if `zat` is matched now. Zat is also used in spoken / written language and I wonder if we should not just match on `zat.`.)_

https://onzetaal.nl/taaladvies/afkortingen-dagen-en-maanden/

@tellarin would you rather have many false positives on `zat` and zo`, or better to have them not in the recognizer and just match on the abbreviation ending with a dot.

**Review by iMicknl — COMMENTED** [review]
Had a quick look. Will do some local testing, but looking good!!

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (iMicknl) [reply Patterns/Dutch/Dutch-DateTime.yaml:120]: @LionbridgeCSII see above. Currently `zat` has a null resolution.

```json
[{
	"text": "zat",
	"start": 0,
	"end": 2,
	"typeName": "datetimeV2.date",
	"resolution": null
}]
```

```json
[{
	"text": "zat.",
	"start": 0,
	"end": 3,
	"typeName": "datetimeV2.date",
	"resolution": {
		"values": [{
			"timex": "XXXX-WXX-6",
			"type": "date",
			"value": "2021-06-05"
		}, {
			"timex": "XXXX-WXX-6",
			"type": "date",
			"value": "2021-06-12"
		}]
	}
}]
```

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (LionbridgeCSII) [reply Patterns/Dutch/Dutch-DateTime.yaml]: added

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (LionbridgeCSII) [reply Patterns/Dutch/Dutch-DateTime.yaml]: fixed as suggested

File: Specs/DateTime/Dutch/DateParser.json
Comment (tellarin) [inline Specs/DateTime/Dutch/DateParser.json:4714]: Is the preposition also tagged in English?

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (tellarin) [inline Specs/DateTime/Dutch/DateTimeModel.json:119]: Interesting bug. The second mask should be XX00.

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (LionbridgeCSII) [reply Patterns/Dutch/Dutch-DateTime.yaml]: fixed resolution of "zat" without dot. 
 Since "zat|zon?|vrij" are ambiguous, they are now extracted (as isolated dates) only if followed by a dot (and not at the end of the sentence) or if preceded by "op|voor". 
Test cases added.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (tellarin) [reply Specs/DateTime/Dutch/DateTimeModel.json:691]: Yes, we should open an issue for this as low-priority.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (tellarin) [inline Specs/DateTime/Dutch/DateTimeModel.json:725]: Here it's different, as "morgen" is specifying either morning and then there should be no 20h resolution; or tomorrow, then the date is wrong.
What is the corresponding output in English and German?

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (tellarin) [inline Specs/DateTime/Dutch/DateTimeModel.json:778]: This is incorrect. As the entity is clearly 8am.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (tellarin) [inline Specs/DateTime/Dutch/DateTimeModel.json:1915]: @iMicknl, what's your opinion on this one?

**Review by tellarin — CHANGES_REQUESTED** [review]

File: Specs/DateTime/Dutch/DateParser.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateParser.json:4714]: yes, in English "Text": "on Sunday of next week".

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:119]: yes, I have not modified it because it appears also in the English model ("Who are us presidents in 90 s."). 
I will try to fix it then.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:725]: "morgen" here should be interpreted as "tomorrow" (the date seems correct because the reference is 2016-11-07). The English counterpart is "I'll go back tomorrow 8:00am" which is resolved to "2016-11-08T08:00"

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:778]: ups, missed this one, will fix it.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:1915]: (note the comment about this case in the PR description)

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (iMicknl) [reply Patterns/Dutch/Dutch-DateTime.yaml]: I would say that abbreviations without a dot like zat/zo/zon/vrij are not commonly used. If I am missing a use-case, we would know soon enough and it shouldn't be too much work. However I would say that the amount of false positives will hurt more if we enable them broadly.

Current solution described by @LionbridgeCSII above sounds good.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (iMicknl) [reply Specs/DateTime/Dutch/DateTimeModel.json:1915]: >"Ik ben 4:00 tot 7 weg" -> "Ik ben weg 4:00 tot 7" (I'll be out 4:00 to 7 oclock). Patters like "4:00 tot 7" that end with a pure number are extracted only if they occur at the end of the sentence.

I am not sure if I understand why it would work at the end of the sentence, but not in the middle. However, it seems like an edge case where you use `4:00 to 7` instead of `4 to 7` or `4:00 to 7:00`.

Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:1915]: I think to avoid extracting false patterns (this behavior is implemented in BaseTimePeriodExtractor where a comment says: "cases like "from 3:30 to 4 people" shuold not be extracted").

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:778]: @iMicknl, these expressions "vanmorgen", "vanavond", "vannacht"... mean "this morning", "this evening", "this night"..., right? Or can they also be used simply for "morning", "evening", "night" in time expressions? 
Because there are a few test cases that seem to imply so, e.g. 
"Ik ben om 7 uur vanavond terug" ("I'll be back afternoon seven pm"), 
"Ik ga terug 20u vanavond, 1 jan" ("I'll go back 8pm in the evening, 1 Jan"). 
But this creates interference between Time and DateTime interpretation:
"20 vanavond" (20 in the evening) -> Time
"20 vanavond" (20 this evening) -> DateTime

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (tellarin) [reply Specs/DateTime/Dutch/DateTimeModel.json:1915]: @LionbridgeCSII, you are right. I forgot this durig the review. nothing to change for now.

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (iMicknl) [reply Specs/DateTime/Dutch/DateTimeModel.json:778]: @LionbridgeCSII The expressions `vanmorgen`, `vanavond`, vannacht` indeed mention the current day. You can't use this expression to define 'morning' etc.

> "Ik ga terug 20u vanavond, 1 jan" ("I'll go back 8pm in the evening, 1 Jan").

I would say that this testcase is incorrect. Or it would be split to `20u vanavond` and `1 jan`.



But this creates interference between Time and DateTime interpretation:
"20 vanavond" (20 in the evening) -> Time
"20 vanavond" (20 this evening) -> DateTime

File: Patterns/Dutch/Dutch-DateTime.yaml
Comment (LionbridgeCSII) [reply Patterns/Dutch/Dutch-DateTime.yaml:20]: removed

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:119]: fixed

File: Specs/DateTime/Dutch/DateTimeModel.json
Comment (LionbridgeCSII) [reply Specs/DateTime/Dutch/DateTimeModel.json:778]: fixed and modified incorrect test cases

**Review by tellarin — APPROVED** [review]
LGTM. Waiting for @iMicknl final review to merge.

**Review by iMicknl — APPROVED** [review]
LGTM! Will do some personal testing in the coming weeks, but look like the specs cover a lot. /cdn.vox-cdn.com/uploads/chorus_asset/file/15979212/Clippy_ShipIt.gif)

**Pull request merged**
